geode-pulse/src/main/java/org/apache/geode/tools/pulse/internal/security/OAuthClientConfig.java

```diff
@@ -0,0 +1,109 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.tools.pulse.internal.security;
+
+import static java.util.Collections.singletonMap;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Profile;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.security.oauth2.client.InMemoryOAuth2AuthorizedClientService;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
+import org.springframework.security.oauth2.client.oidc.web.logout.OidcClientInitiatedLogoutSuccessHandler;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository;
+import org.springframework.security.oauth2.client.web.AuthenticatedPrincipalOAuth2AuthorizedClientRepository;
+import org.springframework.security.oauth2.client.web.OAuth2AuthorizedClientRepository;
+import org.springframework.security.oauth2.core.AuthorizationGrantType;
+
+/**
+ * Configures Pulse to use the OAuth 2 provider defined by properties in {@code pulse.properties}.
+ */
+@Configuration
+@Profile("pulse.authentication.oauth")
+@PropertySource("classpath:pulse.properties")
+public class OAuthClientConfig {
+  @Value("${pulse.oauth.providerId}")
+  private String providerId;
+  @Value("${pulse.oauth.providerName}")
+  private String providerName;
+  @Value("${pulse.oauth.clientId}")
+  private String clientId;
+  @Value("${pulse.oauth.clientSecret}")
+  private String clientSecret;
+  @Value("${pulse.oauth.authorizationUri}")
+  private String authorizationUri;
+  @Value("${pulse.oauth.tokenUri}")
+  private String tokenUri;
+  @Value("${pulse.oauth.userInfoUri}")
+  private String userInfoUri;
+  @Value("${pulse.oauth.jwkSetUri}")
+  private String jwkSetUri;
+  @Value("${pulse.oauth.endSessionEndpoint}")
+  private String endSessionEndpoint;
+  @Value("${pulse.oauth.userNameAttributeName}")
+  private String userNameAttributeName;
+
+  @Bean
+  ClientRegistration clientRegistration() {
+    return ClientRegistration.withRegistrationId(providerId)
+        .authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE)
+        .redirectUriTemplate("{baseUrl}/login/oauth2/code/{registrationId}")
+        .clientId(clientId)
+        .clientSecret(clientSecret)
+        .scope("openid", "CLUSTER:READ", "CLUSTER:WRITE", "DATA:READ", "DATA:WRITE")
+        .authorizationUri(authorizationUri)
+        .tokenUri(tokenUri)
+        .userInfoUri(userInfoUri)
+        .jwkSetUri(jwkSetUri)
+        .providerConfigurationMetadata(
```

Comment:
Is this "end_session_endpoint" provider specific? i.e. it's a strictly UAA thing or others have that as well? Can it be null?

Reply:
That's defined by a draft standard: https://openid.net/specs/openid-connect-session-1_0.html

As far as I can tell, that standard has never been approved, though some OpenID Connect providers do implement it. See for example this Okta documentation: https://developer.okta.com/docs/reference/api/oidc/#well-known-openid-configuration

The sad thing is that UAA does _not_ implement it. But redirecting to UAA's "profile" page has a similar effect. Not exactly what that draft standard defines, but similar.

With the current commit, if you leave the property undefined, Spring assigns the property value expression ("${pulse.oauth.userNameAttributeName}") to the field, and the logout success handler fails on logout.

We did not try supplying null. I'll run a few experiments now. I think the logout success handler will simply skip the "end session" endpoint, and go straight to the "logout success" URL (which we will need to define if we allow the end session endpoint to be null).